### PR TITLE
91 - [Feat] Duplicate Project Names and Web Scraper Fixes

### PR DESCRIPTION
**Pull request by deenasun**
### Description
[//]: # "[Required] Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
Updated `api/webscraper/database.py`
- Wrote `offset_lat_long` function
    - Checks if there are existing projects in Supabase with latitude and longitude coordinates within 0.005 of the current coordinate
    - Offsets latitude and longitude by 0.01 until there are no overlaps
- Added checks to update all other key development milestones if "Start of operations" has been completed
- Created `merge_projects` function
    - Merges all duplicate projects that have matching names
    - Collects all fields from all matching projects, then updates a single project
    - For size, this function adds all the sizes of any matching projects
    - Lastly, the function deletes any duplicate projects from the table so that the table only stores one project

Updated names of milestones
- "Application for permit to ORES" is now "Application for permit"
- "Issuance of permit from ORES" is now "Issuance of permit"

Added new helper functions in `scraper_utils.py`
- `find_keyword` finds the main name for the project name
    - Skips any projects that only have an id or code as their name (i.e. any projects that don't have spaces in their name)
    - For other projects, it returns a substring of the name up until any `*` symbols, numbers, or renewable technology labels such as "Solar" and "Wind"
    - Ex: "Cohocton*" would return a keyword of "Cohocton"
    - Ex: "Cohocton Wind Project" would return a keyword of "Cohocton"
- `combine_projects` combines fields from two projects dicts together by appending/replacing values that are missing
- `pass_all_kdms` marks all key development milestones as `completed` (used if a project's "Start of operations" has been completed)

### Screenshots
[//]: # "Add any screenshots that will help your PR reviewer understand what’s going on in the PR!"
[//]: # "[Required] For frontend tasks: add screenshots of the layout and expected behavior (GIFs if you’re feeling extra fancy)!"
[//]: # "Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."

[image placeholder]: 

## How to review
[//]: # "/[Required] Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"
```
git fetch origin
git checkout deenasun/91-feat-duplicate-project-names-and-web-scraper-fixes
```

To run functions in `database.py`, they can now be triggered by backend endpoint calls through `.../api/[FUNCTION]`

## Next steps
[//]: # "[Optional] - What has not yet been added to the PR, doesn’t yet have any functionality — essentially, what needs to be fixed or added."
[//]: # "[Optional] any temporary fixes in this PR that should be cleaned up later."

### Online sources
[//]: # "[Optional] insert links to any useful tutorials or documentation that you used for this PR!"
- Selecting from Supabase with filters: https://supabase.com/docs/reference/python/using-filters

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @itsliterallymonique

### CODE REVIEW

File: api/webscraper/database.py [hunk 1]
```diff
@@ -742,13 +841,125 @@ def ores_permitted_to_database() -> None:
...
+                )
+
             try:
                 response = supabase.table(supabase_table).insert(project).execute()
                 print("INSERT", response, "\n")
             except Exception as exception:
                 print(exception)
 
 
+def merge_projects() -> None:
```
[the remark this answers is not in the document]
Comment (itsliterallymonique) [reply api/webscraper/database.py:861]: Just so I can check it's functional

**Review by itsliterallymonique — COMMENTED** [review]
Looks good! Just a few questions + comments. Before you merge I just also want to check the current table after running the webscraper. Can you run this + send me the table where I can see the results? Thanks!

File: api/app.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from flask import Flask
```
[the remark this answers is not in the document]
Comment (deenasun) [reply api/app.py:1]: This is for hosting the backend as a server. Flask lets us host all the functions here as URL endpoints. So every time we go to `[websitedomain]/api/[function_name]` the functions below get called

File: api/webscraper/database.py [hunk 3]
```diff
@@ -1,32 +1,39 @@
...
 )
 
+load_dotenv(os.path.join(os.path.dirname(__file__), "../../.env.local"))
+
 url: str = os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
 key: str = os.environ.get("NEXT_PUBLIC_SUPABASE_ANON_KEY")
 supabase: Client = create_client(url, key)
 supabase_table: str = (
-    "Projects_duplicate"  # TODO: modify based on which table in supabase we want to edit
+    "Projects_test"  # TODO: modify based on which table in supabase we want to edit
```
[the remark this answers is not in the document]
Comment (deenasun) [reply api/webscraper/database.py:36]: will do right now!

File: api/webscraper/database.py [hunk 4]
```diff
@@ -498,6 +567,16 @@ def nyiso_to_database_helper(projects, sheet_name):
...
                 )
+
+                # offset latitude and longitude to avoid overlaps
+                if (
+                    project.get("latitude", None) is not None
+                    and project.get("longitude", None) is not None
+                ):
+                    project["latitude"], project["longitude"] = offset_lat_long(
+                        project["latitude"], project["longitude"]
+                    )
```
[the remark this answers is not in the document]
Comment (deenasun) [reply api/webscraper/database.py:578]: offset_lat_long is already a function, @itsliterallymonique can you clarify which lines you want me to make into a function?

File: api/webscraper/database.py [hunk 5]
```diff
@@ -742,13 +841,125 @@ def ores_permitted_to_database() -> None:
...
+                            kdm=update["key_development_milestones"],
+                        )
+                # mark the current duplicate for deletion
+                duplicates_to_delete.append(matching_project["id"])
+
+            start_of_operations = [
+                milestone
+                for milestone in update["key_development_milestones"]
+                if milestone["milestoneTitle"] == "Start of operations"
+            ][0]
```
[the remark this answers is not in the document]
Comment (deenasun) [reply api/webscraper/database.py:925]: these lines find the "start of operations" kdm. I use list iteration to get a list of all kdms whose title matches "Start of operations" (which is just a single one). So then we get the first item from this list, which is the "Start of operations" kdm

File: api/webscraper/utils/scraper_utils.py [hunk 6]
```diff
@@ -133,6 +134,46 @@ def turn_timestamp_to_string(timestamp):
     return timestamp.to_pydatetime().strftime("%Y-%m-%d")
 
 
+def find_keyword(project_name):
+    if " " not in project_name:
```
[the remark this answers is not in the document]
Comment (deenasun) [reply api/webscraper/utils/scraper_utils.py:138]: Yes! so any names that have like "*", numbers, or "wind" and "solar" after the main name

File: api/webscraper/database.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (itsliterallymonique) [reply api/webscraper/database.py:861]: Sounds good! Dataset looks great!!

File: api/webscraper/utils/scraper_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (itsliterallymonique) [reply api/webscraper/utils/scraper_utils.py:138]: okay sounds good. I'll try my best to clarify all the different cases with our NPO asap so we can add any other ones.

[the remark this answers is not in the document]
Comment (itsliterallymonique) [reply api/webscraper/utils/scraper_utils.py:138]: Can you make sure to add this to your documentation? Just so me + future PL can remember!

File: api/webscraper/database.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (itsliterallymonique) [reply api/webscraper/database.py:578]: I think it's fine to not put it as a function. But I noticed this if statement:
```
if (
                project.get("latitude", None) is not None
                and project.get("longitude", None) is not None
            ):
                project["latitude"], project["longitude"] = offset_lat_long(
                    project["latitude"], project["longitude"]
                )
```
is used quite a few times in this file. Is there a way we can make it a function so the code doesn't have repeats?

[the remark this answers is not in the document]
Comment (deenasun) [reply api/webscraper/database.py:578]: yeah I can make this more concise!

**Pull request closed without merging**